File: LeetCode/Note_0049_Group_Anagrams/Note_0049_Group_Anagrams.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.omg.PortableInterceptor.SUCCESSFUL;

class Solution {

    public List<List<String>> groupAnagrams(String[] strs) {
        
        Map<String, List> hash = new HashMap<>();

        for (String s: strs) {
            char[] c = s.toCharArray();
            Arrays.sort(c);
            
            String key = String.valueOf(c);
            System.out.println(key);

            if (!hash.containsKey(key)) {
                hash.put(key, new ArrayList());
            }
            hash.get(key).add(s);
        }
        System.out.println(hash);
        
        return new ArrayList(hash.values());
    }

    public List<List<String>> groupAnagrams_count(String[] strs) {
        int[] count = new int[26];
        Map<String, List> hash = new HashMap<>();
        for (String s: strs) {
            Arrays.fill(count, 0);
            for (char c: s.toCharArray()) {
                count[c - 'a']++;
            }

            StringBuilder sb = new StringBuilder("");
            for (int i = 0; i < count.length; i++) {
                sb.append(count[i]);
                sb.append("#");
            }

            String key = sb.toString();
            if (!hash.containsKey(key)) {
                hash.put(key, new ArrayList<>());
            }
            hash.get(key).add(s);
            System.out.println(hash.toString());
        }

        return new ArrayList(hash.values());
    }


    public static void main(String[] arg) {
        Solution solution = new Solution();
        List<List<String>> result = solution.groupAnagrams(new String[]{"eat","tea","tan","ate","nat","bat"});
        List<List<String>> result_count = solution.groupAnagrams_count(new String[]{"eat","tea","tan","ate","nat","bat"});
        System.out.println(result);
        System.out.println(result_count);
    }

}

File: LeetCode/Note_0049_Group_Anagrams/Note_0049_Group_Anagrams.py
from typing import List

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:

        result = {}
        for s in strs:
            string_s = ' '.join(sorted(s))
            
            if string_s in result:
                result[string_s].append(s)
            else:
                result[string_s] = [s]

        return result.values()

    def groupAnagrams_tuple(self, strs: List[str]) -> List[List[str]]:
        result = {}
        for s in strs:
            string_s = tuple(sorted(s))
            print(string_s)
            if string_s in result:
                result[string_s].append(s)
            else:
                result[string_s] = [s]

        return result.values()

if __name__ == "__main__":
    print("0049")
    solution = Solution()
    
    strs = ["eat","tea","tan","ate","nat","bat"]
    print(solution.groupAnagrams(strs))

    print(solution.groupAnagrams_tuple(strs))
    
# T:O(K * NlogN)
# S:O(K * N)